File: AA3-EV01/frontend/src/app/services/turno.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { Turno, EstadoTurno } from '../models/turno.model';

/**
 * Servicio Angular para la gestión de turnos de la barbería.
 * Consume los endpoints REST del backend Spring Boot.
 *
 * Inyectable en cualquier componente de la aplicación mediante DI.
 *
 * @author BarberTurno Team
 * @version 1.0
 */
@Injectable({
  providedIn: 'root' // Disponible en toda la aplicación (singleton)
})
export class TurnoService {

  /** URL base de la API REST del backend */
  private readonly API_URL = 'http://localhost:8080/api/turnos';

  /**
   * Constructor: inyecta el cliente HTTP de Angular.
   *
   * @param http cliente HTTP para realizar peticiones REST
   */
  constructor(private http: HttpClient) {}

  // ─── Consultas ────────────────────────────────────────────────────────────

  /**
   * Obtiene todos los turnos registrados en el sistema.
   *
   * @returns Observable con la lista de todos los turnos
   */
  obtenerTodos(): Observable<Turno[]> {
    return this.http.get<Turno[]>(this.API_URL).pipe(
      catchError(this.manejarError)
    );
  }

  /**
   * Obtiene un turno específico por su ID.
   *
   * @param id identificador del turno
   * @returns Observable con el turno encontrado
   */
  obtenerPorId(id: number): Observable<Turno> {
    return this.http.get<Turno>(`${this.API_URL}/${id}`).pipe(
      catchError(this.manejarError)
    );
  }

  /**
   * Filtra los turnos por fecha específica.
   *
   * @param fecha fecha en formato yyyy-MM-dd
   * @returns Observable con la lista de turnos de esa fecha
   */
  obtenerPorFecha(fecha: string): Observable<Turno[]> {
    const params = new HttpParams().set('fecha', fecha);
    return this.http.get<Turno[]>(`${this.API_URL}/fecha`, { params }).pipe(
      catchError(this.manejarError)
    );
  }

  /**
   * Filtra los turnos de un barbero en una fecha específica.
   *
   * @param barbero nombre del barbero
   * @param fecha   fecha en formato yyyy-MM-dd
   * @returns Observable con la lista de turnos filtrados
   */
  obtenerPorBarberoYFecha(barbero: string, fecha: string): Observable<Turno[]> {
    const params = new HttpParams()
      .set('barbero', barbero)
      .set('fecha', fecha);
    return this.http.get<Turno[]>(`${this.API_URL}/barbero`, { params }).pipe(
      catchError(this.manejarError)
    );
  }

  /**
   * Busca turnos por nombre de cliente (búsqueda parcial).
   *
   * @param nombre nombre o fragmento del nombre del cliente
   * @returns Observable con la lista de turnos que coinciden
   */
  buscarPorCliente(nombre: string): Observable<Turno[]> {
    const params = new HttpParams().set('nombre', nombre);
    return this.http.get<Turno[]>(`${this.API_URL}/buscar`, { params }).pipe(
      catchError(this.manejarError)
    );
  }

  // ─── Mutaciones ───────────────────────────────────────────────────────────

  /**
   * Crea un nuevo turno en el sistema.
   *
   * @param turno datos del turno a crear
   * @returns Observable con el turno creado (incluye el ID asignado)
   */
  crearTurno(turno: Turno): Observable<Turno> {
    return this.http.post<Turno>(this.API_URL, turno).pipe(
      catchError(this.manejarError)
    );
  }

  /**
   * Actualiza todos los datos de un turno existente.
   *
   * @param id    identificador del turno
   * @param turno nuevos datos del turno
   * @returns Observable con el turno actualizado
   */
  actualizarTurno(id: number, turno: Turno): Observable<Turno> {
    return this.http.put<Turno>(`${this.API_URL}/${id}`, turno).pipe(
      catchError(this.manejarError)
    );
  }

  /**
   * Cambia únicamente el estado de un turno.
   *
   * @param id     identificador del turno
   * @param estado nuevo estado a aplicar
   * @returns Observable con el turno actualizado
   */
  cambiarEstado(id: number, estado: EstadoTurno): Observable<Turno> {
    return this.http.patch<Turno>(
      `${this.API_URL}/${id}/estado`,
      { estado }
    ).pipe(
      catchError(this.manejarError)
    );
  }

  /**
   * Elimina un turno del sistema.
   *
   * @param id identificador del turno a eliminar
   * @returns Observable vacío al completarse
   */
  eliminarTurno(id: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/${id}`).pipe(
      catchError(this.manejarError)
    );
  }

  // ─── Manejo de errores ────────────────────────────────────────────────────

  /**
   * Maneja los errores HTTP de las peticiones a la API.
   * Transforma los errores del backend en mensajes amigables.
   *
   * @param error error HTTP recibido
   * @returns Observable que emite el error transformado
   */
  private manejarError(error: any): Observable<never> {
    let mensajeError = 'Ha ocurrido un error inesperado.';

    if (error.status === 404) {
      mensajeError = 'El turno solicitado no fue encontrado.';
    } else if (error.status === 409) {
      // El backend envía el mensaje de conflicto de horario
      mensajeError = error.error?.error || 'El horario ya está ocupado.';
    } else if (error.status === 400) {
      mensajeError = 'Datos del formulario inválidos. Revise los campos.';
    } else if (error.status === 0) {
      mensajeError = 'No se pudo conectar al servidor. Verifique su conexión.';
    }

    console.error('[TurnoService] Error HTTP:', error);
    return throwError(() => new Error(mensajeError));
  }
}

File: AA3-EV01/backend/src/main/java/com/barberturno/service/TurnoService.java
package com.barberturno.service;

import com.barberturno.dto.TurnoDTO;
import com.barberturno.model.EstadoTurno;
import com.barberturno.model.Turno;
import com.barberturno.repository.TurnoRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

/**
 * Servicio que contiene la lógica de negocio para la gestión de turnos.
 * Actúa como intermediario entre el controlador y el repositorio.
 *
 * @author BarberTurno Team
 * @version 1.0
 */
@Service
@Transactional
public class TurnoService {

    /** Repositorio para acceso a datos de turnos */
    private final TurnoRepository turnoRepository;

    /**
     * Constructor con inyección de dependencias.
     *
     * @param turnoRepository repositorio JPA de turnos
     */
    @Autowired
    public TurnoService(TurnoRepository turnoRepository) {
        this.turnoRepository = turnoRepository;
    }

    // ─── Creación ─────────────────────────────────────────────────────────────

    /**
     * Crea un nuevo turno validando que no exista conflicto de horario.
     *
     * @param dto datos del turno a crear
     * @return el turno creado con su ID asignado
     * @throws IllegalStateException si el horario ya está ocupado
     */
    public Turno crearTurno(TurnoDTO dto) {
        // Verificar disponibilidad del horario para el barbero
        boolean ocupado = turnoRepository.existeTurnoOcupado(
                dto.getBarbero(), dto.getFecha(), dto.getHora());

        if (ocupado) {
            throw new IllegalStateException(
                "El barbero " + dto.getBarbero() +
                " ya tiene un turno reservado el " + dto.getFecha() +
                " a las " + dto.getHora());
        }

        // Mapear DTO a entidad y guardar
        Turno turno = mapearDTOaEntidad(dto);
        return turnoRepository.save(turno);
    }

    // ─── Consultas ────────────────────────────────────────────────────────────

    /**
     * Obtiene todos los turnos registrados en el sistema.
     *
     * @return lista completa de turnos
     */
    @Transactional(readOnly = true)
    public List<Turno> obtenerTodos() {
        return turnoRepository.findAll();
    }

    /**
     * Busca un turno por su ID.
     *
     * @param id identificador del turno
     * @return Optional con el turno si existe
     */
    @Transactional(readOnly = true)
    public Optional<Turno> obtenerPorId(Long id) {
        return turnoRepository.findById(id);
    }

    /**
     * Obtiene todos los turnos de una fecha específica.
     *
     * @param fecha fecha a consultar
     * @return lista de turnos en esa fecha
     */
    @Transactional(readOnly = true)
    public List<Turno> obtenerPorFecha(LocalDate fecha) {
        return turnoRepository.findByFecha(fecha);
    }

    /**
     * Obtiene los turnos de un barbero en una fecha específica.
     *
     * @param barbero nombre del barbero
     * @param fecha   fecha a consultar
     * @return lista de turnos del barbero en esa fecha
     */
    @Transactional(readOnly = true)
    public List<Turno> obtenerPorBarberoYFecha(String barbero, LocalDate fecha) {
        return turnoRepository.findByBarberoAndFecha(barbero, fecha);
    }

    /**
     * Busca turnos por el nombre del cliente (búsqueda parcial).
     *
     * @param nombre nombre o fragmento del nombre del cliente
     * @return lista de turnos que coinciden
     */
    @Transactional(readOnly = true)
    public List<Turno> buscarPorCliente(String nombre) {
        return turnoRepository.findByNombreClienteContainingIgnoreCase(nombre);
    }

    // ─── Actualización ────────────────────────────────────────────────────────

    /**
     * Actualiza los datos de un turno existente.
     *
     * @param id  identificador del turno a actualizar
     * @param dto nuevos datos del turno
     * @return el turno actualizado
     * @throws RuntimeException si el turno no existe
     */
    public Turno actualizarTurno(Long id, TurnoDTO dto) {
        // Verificar existencia del turno
        Turno turnoExistente = turnoRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Turno no encontrado con ID: " + id));

        // Actualizar campos del turno
        turnoExistente.setNombreCliente(dto.getNombreCliente());
        turnoExistente.setTelefonoCliente(dto.getTelefonoCliente());
        turnoExistente.setBarbero(dto.getBarbero());
        turnoExistente.setFecha(dto.getFecha());
        turnoExistente.setHora(dto.getHora());
        turnoExistente.setServicio(dto.getServicio());
        turnoExistente.setObservaciones(dto.getObservaciones());

        // Actualizar estado si viene en el DTO
        if (dto.getEstado() != null) {
            turnoExistente.setEstado(dto.getEstado());
        }

        return turnoRepository.save(turnoExistente);
    }

    /**
     * Cambia únicamente el estado de un turno.
     *
     * @param id     identificador del turno
     * @param estado nuevo estado a aplicar
     * @return el turno con el estado actualizado
     * @throws RuntimeException si el turno no existe
     */
    public Turno cambiarEstado(Long id, EstadoTurno estado) {
        Turno turno = turnoRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Turno no encontrado con ID: " + id));

        turno.setEstado(estado);
        return turnoRepository.save(turno);
    }

    // ─── Eliminación ──────────────────────────────────────────────────────────

    /**
     * Elimina un turno por su ID.
     *
     * @param id identificador del turno a eliminar
     * @throws RuntimeException si el turno no existe
     */
    public void eliminarTurno(Long id) {
        if (!turnoRepository.existsById(id)) {
            throw new RuntimeException("Turno no encontrado con ID: " + id);
        }
        turnoRepository.deleteById(id);
    }

    // ─── Métodos privados de apoyo ────────────────────────────────────────────

    /**
     * Convierte un TurnoDTO a una entidad Turno.
     *
     * @param dto objeto DTO con los datos del turno
     * @return entidad Turno lista para persistir
     */
    private Turno mapearDTOaEntidad(TurnoDTO dto) {
        return new Turno(
            dto.getNombreCliente(),
            dto.getTelefonoCliente(),
            dto.getBarbero(),
            dto.getFecha(),
            dto.getHora(),
            dto.getServicio(),
            dto.getObservaciones()
        );
    }
}
